pkg/ingest/conn.go

```diff
@@ -55,76 +57,128 @@ func HandleConnections(
 		// In either case, the writer is closed.
 		m.register(conn)
 		go func() {
-			h(conn, w, idGen, connectedClients)
+			defer conn.Close()
+			h(reader(conn), w, idGen, connectedClients)
 			w.Stop() // make sure it's flushed
 			m.remove(conn)
 		}()
 	}
 }
 
+// ErrIllegalTopicFormat is returned if a record did not start with a proper topic prefix.
+var ErrIllegalTopicFormat = errors.New("illegal topic formatting")
+
+// RecordReader emits records that are prefixed with a topic.
+// It returns io.EOF if the underlying record source has no more data.
+type RecordReader func() (record []byte, err error)
+
+// NewDynamicRecordReader returns a record reader that expects each input record from r
+// to start with a space-delimited topic identifier.
+func NewDynamicRecordReader(r io.Reader) RecordReader {
+	br := bufio.NewReader(r)
+
+	return func() ([]byte, error) {
+		for {
+			l, err := br.ReadBytes('\n')
+			if err != nil {
+				return nil, err
+			}
+			// Validate that a correct topic prefix exists.
+			if i := bytes.IndexByte(l, ' '); i < 0 || !isValidTopic(l[:i]) {
+				return nil, ErrIllegalTopicFormat
+			}
+			return l, nil
+		}
```

Comment:
Why is this code in a `for` loop?

Reply:
Eh, I think I had a different API before and forgot to drop that part.